src/activity_tracker.cpp

```diff
@@ -0,0 +1,136 @@
+#include "activity_tracker.h"
+
+#include "game_constants.h"
+#include "options.h"
+#include "string_formatter.h"
+
+int activity_tracker::weariness() const
+{
+    if( intake > tracker ) {
+        return tracker * 0.5;
+    }
+    return tracker - intake * 0.5;
+}
+
+// Called every 5 minutes, when activity level is logged
+void activity_tracker::try_reduce_weariness( int bmr, bool sleeping )
+{
+    tick_counter++;
+    if( average_activity() - NO_EXERCISE <= std::numeric_limits<float>::epsilon() ) {
+        low_activity_ticks++;
+        // Recover twice as fast at rest
+        if( sleeping ) {
+            low_activity_ticks++;
+        }
+    }
+
+    const float recovery_mult = get_option<float>( "WEARY_RECOVERY_MULT" );
+
+    if( low_activity_ticks >= 6 ) {
+        int reduction = tracker;
+        // 1/20 of whichever's bigger
+        if( bmr > reduction ) {
+            reduction = bmr * recovery_mult;
+        } else {
+            reduction *= recovery_mult;
+        }
+        low_activity_ticks -= 6;
+
+        tracker -= reduction;
+    }
+
+    if( tick_counter >= 12 ) {
+        intake *= 1 - recovery_mult;
+        tick_counter -= 12;
+    }
+
+    // Normalize values, make sure we stay above 0
+    intake = std::max( intake, 0 );
+    tracker = std::max( tracker, 0 );
+    tick_counter = std::max( tick_counter, 0 );
+    low_activity_ticks = std::max( low_activity_ticks, 0 );
+}
+
+void activity_tracker::weary_clear()
+{
+    tracker = 0;
+    intake = 0;
+    low_activity_ticks = 0;
+    tick_counter = 0;
+}
+
+std::string activity_tracker::debug_weary_info() const
```

Comment:
Good idea! Note that this information would also be useful in weary_tests.cpp (tracked with each change of weariness level, as weariness and weary_threshold are now).